### PR TITLE
docs: The rust part...

### PR DESCRIPTION
**Pull request by hunger**

### CODE REVIEW

File: api/rs/slint/mcu.md [hunk 1]
```diff
@@ -13,36 +13,26 @@ Writing an application in Rust that runs on a MCU requires several prerequisites
...
-crates, tools and training materials. These resources should guide you through the initial setup and often come with "hello world" examples
+and the curated list of [Awesome Embedded Rust](https://github.com/rust-embedded/awesome-embedded-rust) for a wide range of
+crates, tools and training materials. These resources should guide you through the initial setup. Many include a "hello world" example
 to get started with your device.
 
-Slint requires a global memory allocator. That is currently only possible in the nightly version of Rust, because the support for using a custom global
-allocator in a bare metal environment with `#![no_std]` has not been stabilized yet (see [#51540](https://github.com/rust-lang/rust/issues/51540) or
-[#66741](https://github.com/rust-lang/rust/issues/66741) for tracking issues).
+Slint requires a custom global memory allocator in a bare metal environment with `#![no_std]`.
+The rust team hasn't yet stabilized this feature, so only the nightly version
```
Comment (ogoffart) [inline api/rs/slint/mcu.md:21]: Note that this is stabilized in the upcomming 1.68   (one of the tracking issue is closed)

File: api/rs/slint/mcu.md [hunk 2]
```diff
@@ -13,36 +13,26 @@ Writing an application in Rust that runs on a MCU requires several prerequisites
...
-
-[dependencies.slint]
-version = "0.3.4"
-default-features = false
-features = ["compat-0-3-0", "unsafe-single-threaded", "libm"]
-
-[build-dependencies]
-slint-build = "0.3.4"
+```sh
+cargo add slint@1.0.0 --no-default-features --features "compat-1-0-0 unsafe-single-threaded libm"
```
Comment (ogoffart) [inline api/rs/slint/mcu.md:35]: i have mixed feelings about that. I think it is usefull to show the Cargo.toml
Maybe we can have both.

Note that this should say that the use of edition 2021 is mandatory.
Also it is missing slint-build 
Note that the other PR calls the `compat-1-0`  without the last 0

Comment (tronical) [reply api/rs/slint/mcu.md:35]: I agree. I see a future where editing `Cargo.toml` will not be necessary anymore, but for the time being it's still the common thing.

Another argument: That `cargo add` line is great for a one-time use, when you already know what features you need. But after that as a new user I'd be lost. "Ok, they have this magic line of features, but how do I add more after that?". Showing the structure of `Cargo.toml` solves that, because then the user knows where to make further edits.

I'm strongly in favor of keeping `Cargo.toml` for now.

File: api/rs/slint/mcu.md [hunk 3]
```diff
@@ -55,12 +45,18 @@ In the snippet above, three features are selected:
...
 
 It might be necessary to enable the [Feature resolver version 2](https://doc.rust-lang.org/cargo/reference/features.html#feature-resolver-version-2)
-in your Cargo.toml if you notice that your dependencies are attempting to build with `std` support even if it was disabled.
+in your Cargo.toml if you notice that your dependencies are attempting to build with `std` support even when disabled.
 This is the default when using the Rust 2021 Edition or later.
 
+Then add the `slint-build` crate as a build dependency:
+
+```sh
+cargo add slint-build@1.0.0 --build
```
Comment (tronical) [inline api/rs/slint/mcu.md:57]: I'm curious, is there a specific reason to pass `--build` as the last argument instead of after `add`?

File: api/rs/slint/mcu.md [hunk 4]
```diff
@@ -80,42 +76,42 @@ in a format that's suitable for the software based renderer we're going to use.
...
 
-Typically, a graphical application in hosted environments is comprised of at least three different tasks:
+Typically, a graphical application in hosted environments has at least three different tasks:
 
- * Receives user input from operation system APIs.
- * Reacts to the input by performing application specific computations.
- * Renders an updated user interface and presents it on the screen using device-independent operating system APIs.
+ * Receive user input from operation system APIs.
+ * React to the input by performing application specific computations.
+ * Render an updated user interface and presents it on the screen using device-independent operating system APIs.
```
Comment (tronical) [inline api/rs/slint/mcu.md:83]: I remember discussing this with Kavindra when we wrote it and I also started with "Receive", "React", and "Render" and she pointed out that the tense should be consistent by using "receives", "reacts", and "renders". It makes sense to me to preserve the present tense of "has". Why did you change it to infinite?

File: api/rs/slint/mcu.md [hunk 5]
```diff
@@ -80,42 +76,42 @@ in a format that's suitable for the software based renderer we're going to use.
...
+task of receiving user input and forwarding it to the user interface layer, and rendering the user interface to the screen.
 
-In bare metal environments it is your responsibility to substitute and connect functionality that is otherwise provided by the operating system:
+In bare metal environments it's your responsibility to substitute and connect functionality that's otherwise provided by the operating system:
 
- * Select crates that allow you to initialize the chips that operate peripherals, such as a touch input or display controller.
-   If there are no crates, you may have to to develop your own drivers.
- * Drive the event loop yourself by querying peripherals for input, forwarding that input into computational modules of your
+ * Select crates that initialize the chips that operate peripherals, such as a touch input or display controller.
+   If there are no crates, you may have to develop your own drivers.
```
Comment (tronical) [inline api/rs/slint/mcu.md:91]: I prefer very much the old wording, where the act of selecting enables the developer to initialise the chips. This new phrase reads as if the act of selecting ends up initializing them.

File: api/rs/slint/mcu.md [hunk 6]
```diff
@@ -80,42 +76,42 @@ in a format that's suitable for the software based renderer we're going to use.
...
+ * Drive the event loop by querying peripherals for input, forwarding that input into computational modules of your
```
Comment (tronical) [inline api/rs/slint/mcu.md:92]: Why did you remove the "yourself" part? I think that's quite essential in terms of conveying responsibility.

File: api/rs/slint/mcu.md [hunk 7]
```diff
@@ -80,42 +76,42 @@ in a format that's suitable for the software based renderer we're going to use.
...
    application and instructing Slint to render the user interface.
 
-In Slint, the two primary APIs you need to use to accomplish these tasks are the [`slint::platform::Platform`] trait as well as the [`slint::Window`] struct.
+In Slint, the two primary APIs you need are the [`slint::platform::Platform`] trait and the [`slint::Window`] struct.
```
Comment (tronical) [inline api/rs/slint/mcu.md:95]: When I wrote it, the reference to the tasks in the intro was deliberate, to guide the developer. Why did you remove the reference?

File: api/rs/slint/mcu.md [hunk 8]
```diff
@@ -80,42 +76,42 @@ in a format that's suitable for the software based renderer we're going to use.
...
 In the following sections we're going to cover how to use them and how they integrate into your event loop.
 
 ### The `Platform` Trait
 
 The [`slint::platform::Platform`] trait defines the interface between Slint and platform APIs typically provided by operating and windowing systems.
 
-You need to provide a minimal implementation of this trait and call [`slint::platform::set_platform`] before constructing your Slint application.
+You need to write a minimal implementation of this trait and call [`slint::platform::set_platform`] before constructing your Slint application.
```
Comment (tronical) [inline api/rs/slint/mcu.md:102]: Somebody else could have written the implementation (like when using the mcu-board-support create). I think "provide" is the correct in the context of the API contract here.

File: api/rs/slint/mcu.md [hunk 9]
```diff
@@ -80,42 +76,42 @@ in a format that's suitable for the software based renderer we're going to use.
...
 
 This minimal implementation needs to cover two functions:
 
- * `fn create_window_adapter(&self) -> Result<Rc<dyn WindowAdapter + 'static>, PlatformError>;`: Implement this function to return an implementation of the `WindowAdapter`
-   trait that will be associated with the Slint components you create. We provide a convenience struct [`slint::platform::software_renderer::MinimalSoftwareWindow`]
+ * `fn create_window_adapter(&self) -> Result<Rc<dyn WindowAdapter + 'static>, PlatformError>;`: This function returns an implementation of the `WindowAdapter`
+   trait that will be associated with the Slint components you create. Slint has a convenience struct [`slint::platform::software_renderer::MinimalSoftwareWindow`]
```
Comment (tronical) [inline api/rs/slint/mcu.md:107]: Why did you change this from actionable ("Implement this") to passive? The context here is a to instruct the developer what they have to do. In API documentation of a function that the user can call, I'd use the term "This function returns XYZ", but the context here is a trait where a function needs to be implemented. The user will never call `create_window_adapter`.

File: api/rs/slint/mcu.md [hunk 10]
```diff
@@ -80,42 +76,42 @@ in a format that's suitable for the software based renderer we're going to use.
...
    that implements this trait.
  * `fn duration_since_start(&self) -> Duration`: For animations in `.slint` design files to change properties correctly, Slint needs to know
-   how much time has elapsed between two rendered frames. In a bare metal environment you need to provide a source of time. Often the HAL crate of your
-   device provides a system timer API for this, which you can query in your impementation.
+   how much time has elapsed between two rendered frames. In a bare metal environment you need to have a source of time. Often the HAL crate of your
+   device includes a system timer API, which you can query in your implementation.
```
Comment (tronical) [inline api/rs/slint/mcu.md:111]: Here I also prefer the previous way it was written: The context is the trait implementation where the developer needs to provide something. So starting with the actionable part ("you need to provide") and then providing advice where there may (or may not) be help ("HAL crate provides") is better IMO than saying that you need to "have" a source of time and that in your implementation you can query some other create. The emphasis here should IMO be on the implementation returning something (hence "provide").

File: api/rs/slint/mcu.md [hunk 11]
```diff
@@ -80,42 +76,42 @@ in a format that's suitable for the software based renderer we're going to use.
...
 
-There are additional functions in the trait that you can implement, for example to handle debug output, to delegate the event loop or to implement
-the interface to safely deliver events in multi-threaded environments.
+You may override more functions of this trait, for example to handle debug output, to delegate the event loop,
+or to deliver events in multi-threaded environments.
```
Comment (tronical) [inline api/rs/slint/mcu.md:114]: "Overriding" is a good choice here, indeed better than before and consistent with the Rust language docs. 👍

File: api/rs/slint/mcu.md [hunk 12]
```diff
@@ -183,13 +179,13 @@ With a `Platform` in place, you can write the main event loop to drive all the d
...
-   event loop in a way similar to desktop platforms, using the [`run()`](slint::ComponentHandle::run) function
-   of your component, or use [`slint::run_event_loop()`]. Both of these functions will call your implementation
-   of [`slint::platform::Platform::run_event_loop`].
- * Implement a `loop { ... }` directly in your main function: This is called a super loop architecture and common
-   for programs running in bare metal environments on MCUs. It allows you to initialize you device peripherals
-   and access them without the need to move them into your `Platform` implementation.
+ * Use [`slint::platform::Platform::run_event_loop`] if you want to start the
+   event loop in a way similar to desktop platforms. The
+   [`run()`](slint::ComponentHandle::run) function of your component, and [`slint::run_event_loop()`]
+   will both call [`slint::platform::Platform::run_event_loop`].
```
Comment (tronical) [inline api/rs/slint/mcu.md:185]: I disagree again. The function "run_event_loop" is not *used* by the developer. It needs to be implemented. As the docs say, it is *used* (called) by `run()`.

File: api/rs/slint/mcu.md [hunk 13]
```diff
@@ -183,13 +179,13 @@ With a `Platform` in place, you can write the main event loop to drive all the d
...
+ * Put a `loop { ... }` into your main function: This so called super loop architecture is common
+   for programs running in bare metal environments on MCUs. It allows to initialize device peripherals
+   and access them without the need to move them into the `Platform` implementation.
```
Comment (tronical) [inline api/rs/slint/mcu.md:188]: I prefer the consistent addressing of the developer with "you". We use it all the time, most recently at the beginning of the section: "You can choose between two options". Why remove it?

File: api/rs/slint/mcu.md [hunk 14]
```diff
@@ -233,28 +229,33 @@ loop {
...
-`slint::platform::WindowAdapter` trait and also holds an instance of a [`slint::platform::software_renderer::SoftwareRenderer`]. You obtain access to it
+In the earlier example, we've instantiated a [`slint::platform::software_renderer::MinimalSoftwareWindow`]. This struct implements the
+`slint::platform::WindowAdapter` trait and also holds an instance of a [`slint::platform::software_renderer::SoftwareRenderer`]. You access it
 through the callback parameter of the [`draw_if_needed()`](MinimalSoftwareWindow::draw_if_needed) function.
 
-Depending on the amount of RAM your MCU is equipped with, and the kind of screen that is attached, you can choose between two different ways of using the renderer:
+Depending on the amount of RAM your MCU has, and the kind of screen attached, you can choose between two different ways of using the renderer:
 
- * Use the [`SoftwareRenderer::render()`] function if you have enough RAM to allocate one, or even two, copies of the entire screen (also known as
+ * Use the [`SoftwareRenderer::render()`] function if you have enough RAM to hold one, or even two, copies of the entire screen (also known as
```
Comment (tronical) [inline api/rs/slint/mcu.md:241]: Why did you replace "allocate" with "hold"?

I think "allocation" in the context of "RAM" is idiomatic, while "holding" isn't.

File: api/rs/slint/mcu.md [hunk 15]
```diff
@@ -233,28 +229,33 @@ loop {
...
    frame buffer).
- * Use the [`SoftwareRenderer::render_by_line()`] function to render the entire user interface line by line and send each line of pixels to the screen,
-   typically via the SPI. This requires allocating at least enough RAM to store one single line of pixels.
+ * Use the [`SoftwareRenderer::render_by_line()`] function to render the entire user interface line
+   by line. This requires enough RAM to hold an entire line of pixels in memory.
+   After rendering a line, its data is transfered to the screen, typically via
+   the SPI.
```
Comment (tronical) [inline api/rs/slint/mcu.md:246]: I prefer the previous formulation, where lines are rendered and each line of pixels is sent to the screen via SPI, instead of the the new abstract term "line data" that's sent.

Also, the previous two sentences first (1) introduced the order of the algorithm and (2) then provided a note about RAM requirements. Now this is mixed up: (1) Everything is rendered line by line (2) RAM requirements come in and (3) we're back to the order of things.

File: api/rs/slint/mcu.md [hunk 16]
```diff
@@ -233,28 +229,33 @@ loop {
...
+For convenience, Slint provides an implementation for [`slint::Rgb8Pixel`] and [`slint::platform::software_renderer::Rgb565Pixel`].
 
-#### Rendering into a Buffer
+#### Rendering Into a Buffer
 
-The following example uses double buffering and swaps between them. This requires a graphics driver that can be provided
-with the address of what should be the currently displayed frame buffer, also known as front buffer. A dedicated chip is then responsible for
-reading from RAM and transferring the contents to the attached screen, without any interference of the CPU. Meanwhile, Slint can render into
-the second buffer, the back buffer.
+The following example uses double buffering and swaps between two buffers. Thisi
```
Comment (tronical) [inline api/rs/slint/mcu.md:253]:
```suggestion
The following example uses double buffering and swaps between two buffers. This
```

File: api/rs/slint/mcu.md [hunk 17]
```diff
@@ -387,9 +388,9 @@ loop {
 ```
 
 Note: In our experience, using the synchronous `DrawTarget::fill_contiguous` function is slow. If
-your device is capable of using DMA, you may be able to achieve better performance by using
-two line buffers: One buffer to render into with the CPU, while the other buffer is transferred to
-the screen using DMA asynchronously.
+your device is capable of DMA, you can improve performance by switching between
+two line buffers. Use one buffer to render into using the CPU. The other buffer
+can transfer data to the screen using DMA in the meantime.
```
Comment (tronical) [inline api/rs/slint/mcu.md:393]: This isn't correct I think. The buffer is not the entity that transfers data to the screen. It is the buffer that *is* transferred (passive) using DMA. I'd also keep the asynchronously part since that is crucial to the advertised performance improvement.

**Review by tronical — COMMENTED** [review]
Lots of simplifications that I like, but also a few things I strongly disagree with.

File: api/cpp/docs/generated_code.md [hunk 18]
```diff
@@ -44,9 +44,9 @@ by a `std::vector<T>`.
 Let's assume we've this code in our `.slint` file:
 
 ```slint,no-preview
-SampleComponent := Window {
-    property<int> counter;
-    property<string> user_name;
+component SampleComponent inherits Window {
+    in-oit property<int> counter;
+    out property<string> user_name;
```
[the remark this answers is not in the document]
Comment (ogoffart) [reply api/cpp/docs/generated_code.md]: I do think it makes totally sense to show the out property, and i thought Tobias's idea to have user_name as an out parameter is not too far fetched.

> Even such a "user name entered" property would probably have a value pre-filled for a login screen for example if this is the second time the user logs in.

True, but maybe there isn't.  And it is more appropriate to show what Slint generates that be accurate on the meaning of `user_name`.  Would it be better to rename that property `sample_output_property`  I don't think so.  Maybe `selected-user-name` ?

File: api/rs/slint/mcu.md [hunk 19]
```diff
@@ -13,36 +13,29 @@ Writing an application in Rust that runs on a MCU requires several prerequisites
...
-crates, tools and training materials. These resources should guide you through the initial setup and often come with "hello world" examples
+and the curated list of [Awesome Embedded Rust](https://github.com/rust-embedded/awesome-embedded-rust) for a wide range of
+crates, tools, and training materials. These resources should guide you through the initial setup. Many include a "hello world" example
 to get started with your device.
 
-Slint requires a global memory allocator. That is currently only possible in the nightly version of Rust, because the support for using a custom global
-allocator in a bare metal environment with `#![no_std]` has not been stabilized yet (see [#51540](https://github.com/rust-lang/rust/issues/51540) or
-[#66741](https://github.com/rust-lang/rust/issues/66741) for tracking issues).
+Slint requires a custom global memory allocator in a bare metal environment with `#![no_std]`.
+The rust team hasn't yet stabilized this feature, so only the nightly version
```
Comment (tronical) [inline api/rs/slint/mcu.md:21]:
```suggestion
The Rust team hasn't yet stabilized this feature, so only the nightly version
```

File: api/rs/slint/mcu.md [hunk 20]
```diff
@@ -233,28 +252,31 @@ loop {
...
+Slint provides a SoftwareRenderer for this task.
 
-In the previous example, we've instantiated a [`slint::platform::software_renderer::MinimalSoftwareWindow`]. This struct implements the
-`slint::platform::WindowAdapter` trait and also holds an instance of a [`slint::platform::software_renderer::SoftwareRenderer`]. You obtain access to it
+In the earlier example, we've instantiated a [`slint::platform::software_renderer::MinimalSoftwareWindow`]. This struct implements the
+`slint::platform::WindowAdapter` trait and also holds an instance of a [`slint::platform::software_renderer::SoftwareRenderer`]. You access it
 through the callback parameter of the [`draw_if_needed()`](MinimalSoftwareWindow::draw_if_needed) function.
-
-Depending on the amount of RAM your MCU is equipped with, and the kind of screen that is attached, you can choose between two different ways of using the renderer:
+ * Use the [`SoftwareRenderer::render()`] function if you have enough RAM to hold one, or even two, copies of the entire screen (also known as
```
Comment (tronical) [inline api/rs/slint/mcu.md:261]: This seems like a copy & paste mistake :-)

```suggestion
```

**Review by tronical — APPROVED** [review]
Looks good to me! One small copy & paste mistake inside though

**Review by tronical — APPROVED** [review]
